File: 알고리즘/정지관/javascript/2805.js
let fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().split('\n');
let [n, m] = input[0].split(' ').map(Number)
let arr = input[1].split(' ').map(Number).sort((a, b) => b - a)

let start = 0
let end = arr.reduce((a, b) => Math.max(a, b));

let result = 0;

while(start <= end) {
    let total = 0;
    let mid = parseInt((start + end) / 2)
    for(let x of arr) {
        if(mid < x) total += x - mid 
    }
    if(total < m) {
        end = mid - 1;
    } else {
        result = mid;
        start = mid + 1;
    }
    
}


console.log(result)

File: 알고리즘/유찬홍/백준/02000/2805.c
#include <stdio.h>

int main() {
    int n, m, arr[1000000], max = 0;
    scanf("%d %d", &n, &m);
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
        if (max < arr[i]) max = arr[i];
    }
    int j = 0, k = max - 1, rs;
    while (j <= k) {
        int mid = (j + k) / 2;
        long long int l = 0;
        for (int i = 0; i < n; i++) {
            if (arr[i] > mid) l += arr[i] - mid;
        }
        if (l < m) k = mid - 1;
        else {
            rs = mid;
            j = mid + 1;
        }
    }
    printf("%d", rs);
    return 0;
}